### PR TITLE
feat: add DNS interceptor

### LINKED ISSUE #3350: dns round-robin interceptor + cache
```js
class Handler {
  // 1. resolve dns entries
  // 2. pick an ip for entry list
  // 3. const origin = new URL(opts.origin)
  // 4. origin.hostname = ip
  // 3. call dispatch({ ...opts, origin }, handler)
}

export const dnsInterceptor = dispatch => (opts, handler) => isDNSName(opts.origin.hostname)
  ? dispatch(opts, new Handler(opts, { dispatch, handler }))
  : dispatch(opts, handler)
```

### PR DESCRIPTION
**Pull request by metcoder95**
Closes #3350 

**WIP**

- [x] TTL handling
- [x] Limit max items cached
- [x] Housekeeping: Improve default distribution handling
- [x] Housekeeping: Tests

### CODE REVIEW

**Comment by metcoder95** [comment]
It is on a super early state, but should be ok to review @ronag 

>Please ignore my mess on the distribution and other boilerplate code, I'll improve that later

**Review by ronag — COMMENTED** [review]
We could maybe use https://www.npmjs.com/package/lru-cache for lru?

File: lib/interceptor/dns.js [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+          if (record.family === 4) {
+            acc[4].ips.push({ address: record.address, ttl: record.ttl })
+          } else {
+            acc[6].ips.push({ address: record.address, ttl: record.ttl })
+          }
+
+          return acc
+        },
+        { 4: { ips: [] }, 6: { ips: [] } }
+      )
```
Comment (ronag) [inline lib/interceptor/dns.js:48]: No reduce please. Makes it more diffcult to reason avbout

**Comment by ronag** [comment]
I'm super busy this month. Will do my best to provide feedback. But for now I mostly have time for high level feedback. LGTM so far.

**Comment by metcoder95** [comment]
We can improve the caching with a self made LRU if we want, but I'd rather explore that in further PRs

File: lib/interceptor/dns.js [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+      this.dualStack === false
+        ? records[this.affinity]
+        : records[affinity] ??
+          records[(this.lastIpFamily = (newOffset & 1) === 1 ? 4 : 6)]
+
+    if (family == null) {
+      return this.pick(
+        origin,
+        hostnameRecords,
+        affinity ?? this.lastIpFamily === 4 ? 6 : 4
```
Comment (metcoder95) [inline lib/interceptor/dns.js:73]: fix: respect `dualStack`

File: lib/interceptor/dns.js [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+    if (family.offset === maxInt) {
+      family.offset = 0
+    } else {
+      family.offset++
+    }
+
+    const ip = family.ips[family.offset % family.ips.length]
+
+    const timestamp = Date.now()
+    if (ip.timestamp != null && timestamp - ip.timestamp > ip.ttl * 1000) {
```
Comment (metcoder95) [inline lib/interceptor/dns.js:89]: fix: standardize TTL to ms

File: lib/interceptor/dns.js [hunk 4]
```diff
@@ -256,13 +269,21 @@ class DNSDispatchHandler extends DecoratorHandler {
...
     throw new InvalidArgumentError('Invalid maxTTL. Must be a positive number')
   }
 
+  if (
+    interceptorOpts?.maxItems != null &&
+    (typeof interceptorOpts?.maxItems !== 'number' ||
+      interceptorOpts?.maxItems < 1)
+  ) {
+    throw new InvalidArgumentError(
+      'Invalid maxItems. Must be a positive number'
```
Comment (Uzlopak) [inline lib/interceptor/dns.js:283]: 0 is a positive number

Comment (metcoder95) [reply lib/interceptor/dns.js:283]:
```suggestion
      'Invalid maxItems. Must be a positive number and greater than zero'
```

**Comment by mcollina** [comment]
CI does not seem happy

**Comment by metcoder95** [comment]
I'm intrigued about the Ubuntu ones, but the Window's ones seem odd. I'll check them throughout the week

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by metcoder95** [comment]
are we good to merge?

**Review by ronag — APPROVED** [review]

**Pull request merged**